pkg/backends/calico/routes_test.go

```diff
@@ -0,0 +1,135 @@
+package calico
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func addEndpointSubset(ep *v1.Endpoints, nodename string) {
+	ep.Subsets = append(ep.Subsets, v1.EndpointSubset{
+		Addresses: []v1.EndpointAddress{
+			v1.EndpointAddress{
+				NodeName: &nodename}}})
+}
+
+func buildSimpleService() (svc *v1.Service, ep *v1.Endpoints) {
+	meta := metav1.ObjectMeta{Namespace: "foo", Name: "bar"}
+	svc = &v1.Service{
+		ObjectMeta: meta,
+		Spec: v1.ServiceSpec{
+			Type:                  v1.ServiceTypeClusterIP,
+			ClusterIP:             "127.0.0.1",
+			ExternalTrafficPolicy: v1.ServiceExternalTrafficPolicyTypeLocal,
+		}}
+	ep = &v1.Endpoints{
+		ObjectMeta: meta,
+	}
+	return
+}
+
+var _ = Describe("RouteGenerator", func() {
+	var rg *routeGenerator
+	BeforeEach(func() {
+		rg = &routeGenerator{
+			nodeName:    "foobar",
+			svcIndexer:  cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			epIndexer:   cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			svcRouteMap: make(map[string]string),
+			client: &client{
+				cache: make(map[string]string),
+			},
+		}
+	})
+	Describe("getServiceForEndpoints", func() {
+		It("should get corresponding service for endpoints", func() {
+			// getServiceForEndpoints
+			svc, ep := buildSimpleService()
+			rg.svcIndexer.Add(svc)
+			fetchedSvc, key := rg.getServiceForEndpoints(ep)
+			Expect(fetchedSvc.ObjectMeta).To(Equal(svc.ObjectMeta))
+			Expect(key).To(Equal("foo/bar"))
+		})
+	})
+	Describe("getEndpointsForService", func() {
+		It("should get corresponding endpoints for service", func() {
+			// getEndpointsForService
+			svc, ep := buildSimpleService()
+			rg.epIndexer.Add(ep)
+			fetchedEp, key := rg.getEndpointsForService(svc)
+			Expect(fetchedEp.ObjectMeta).To(Equal(ep.ObjectMeta))
+			Expect(key).To(Equal("foo/bar"))
+		})
+	})
+
+	Describe("(un)setRouteForSvc", func() {
+		svc, ep := buildSimpleService()
```

Comment:
Is this what you want?

This will build an ep object once, and then for each test add a new subset to it, accumulating subsets for each test.